File: lib/day1.ex
defmodule Day1 do
  def part1(lines) do
    lines
    |> Enum.map(&match_digits/1)
    |> Enum.map(&extract/1)
    |> Enum.sum()
  end

  def part2(lines) do
    mapping =
      ~W[one two three four five six seven eight nine]
      |> Enum.zip(Enum.map(1..9, &to_string/1))
      |> Map.new()

    lines
    |> Stream.map(&match_numbers/1)
    |> Stream.map(&extract(&1, mapping))
    |> Enum.sum()
  end

  defp match_numbers(line) do
    first =
      ~r"\d|one|two|three|four|five|six|seven|eight|nine"
      |> Regex.run(line)

    last =
      ~r"\d|eno|enin|thgie|neves|xis|evif|ruof|eerht|owt"
      |> Regex.run(String.reverse(line))
      |> Enum.map(&String.reverse/1)

    [first, last]
  end

  defp match_digits(line) do
    ~r"\d"
    |> Regex.scan(line)
  end

  defp extract(list) do
    ((List.first(list) |> hd) <> (List.last(list) |> hd))
    |> String.to_integer()
  end

  defp extract(list, mapping) do
    first = List.first(list) |> hd
    last = List.last(list) |> hd

    (Map.get(mapping, first, first) <> Map.get(mapping, last, last))
    |> String.to_integer()
  end
end

File: lib/day1.py
import re

WORDS = "one,two,three,four,five,six,seven,eight,nine".split(",")
WORD_TO_NUMBER = dict(zip(WORDS, map(str, range(1, 10))))


def convert_to_number(line, regex):
    matches = regex.findall(line)
    first = matches[0]
    last = matches[-1]
    return int(WORD_TO_NUMBER.get(first, first) + WORD_TO_NUMBER.get(last, last))


def parts(lines, regex):
    return sum(convert_to_number(line, regex) for line in lines)


def main():
    digit_match = re.compile(r"(\d)")
    number_match = re.compile(rf"(?=(\d|{'|'.join(WORDS)}))")
    with open("inputs/day1.txt") as f:
        lines = f.readlines()
    print("part 1:", parts(lines, digit_match))
    print("part 2:", parts(lines, number_match))


if __name__ == "__main__":
    main()
